pc2ccs/pc2v9
# I75 wti scoreboard

### Description of what the PR does

Adds a scoreboard to the WTI team display.  The scoreboard is cached between changes, and dynamically updates whenever any score-affecting change in the contest occurs.  The scoreboard implementation comprises two main parts:  
* Updates to the WTI-API server to support a new **/scoreboard** endpoint which can be used by web clients to request updated scoreboard information, plus listeners in the WTI-API for PC2 score-changing events.
* A new **scoreboard page** component in the WTI-UI Angular code on the client side which displays current standings while listening for updates.   

The WTI-API server listens for PC2 scoring changes, and when a change callback occurs it sends a websocket message to the Angular client-side scoreboard code notifying it that a change has occurred.

Both the WTI-API and WTI-UI code maintain cached copies of the standings and only update them when (1) a notification of a scoring change has been received, AND (2) someone requests the current standings.  In the case of the WTI-UI (Angular client-side) code, updates are only processed when the scoreboard is actually visible, to avoid repeatedly responding to events which aren't currently being viewed by anyone.

The code in the PR passes all PC2v9 and WTI JUnit tests. The code shows no errors in Eclipse, and no warnings that were not previously present.

### Issue which the PR fixes
Fixes #75 

### Environment in which the PR was developed (OS,IDE, Java version, etc.)
Win 10, Eclipse 2019-12, Java 8_201

### Steps to test this PR:

* It might be helpful to first read the documentation at https://github.com/pc2ccs/pc2v9/wiki/WTI-API-Scoreboard-Implementation and https://github.com/pc2ccs/pc2v9/wiki/WTI-UI-Scoreboard-Implementation

* Download the complete code for the PR (both the PC2v9 system and the WTI subproject).

* Execute pc2v9/package.xml to build a complete distribution.

* Copy the pc2-9.7build distribution zip file from the **pc2v9/dist** folder to any convenient location and unzip it.

* Start a PC2 server from the unzipped distribution

* Start a PC2 Admin from the unzipped distribution

* Configure a contest with at least the following:
  * 2 teams
  * 2 problems 
  * 1 AJ account configured to judge both contest problems
  * TWO scoreboard accounts:  "scoreboard1" and "scoreboard2"
The "SumitHello" sample contest will work for all the above EXCEPT that it needs the following changes:
    * Edit the "Hello" problem to expect no input data.
    * Edit the "Sumit" problem to read from "stdin"
    * Add a "scoreboard2" account

* Start an AJ which is configured to judge both problems.

* Start a PC2 board and login as "scoreboard1".  (Do NOT use "scoreboard2" for this login.)

* Use the PC2 Admin to _change the scoreboard2 account password_ (to anything other than "scoreboard2"). (This is to test that the WTI project correctly aborts when it cannot login to the scoreboard account).

* Use the PC2 Admin to start the contest

* Go to the  **/projects** folder in the unzipped distribution and unzip **WebTeamInterfce-1.1.xx.zip** to any convenient location (except not into the pc2 "jake" folder, because the pc2v9.ini files will conflict).

* Go in to the resulting WebTeamInterface-1.1 folder and execute "./bin/pc2wti" to start the WTI-API server.
  * This should cause the WTI-API server to abort with a message on the console.  This is because an earlier step (above) changed the "scoreboard2" account password from the default value ("scoreboard2").

* Use the PC2 Admin to change the scoreboard2 account password to "scoreboard2".

* In the WebTeamInterface folder, again execute the command "./bin/pc2wti".  This should successfully start up the WTI Jetty server.

* Open a web browser to http://localhost:8080.  This should display the WTI "Login" screen.

* Enter the credentials for a valid team account, press "Log In".  This should display the WTI main UI, defaulting to the "Runs" screen.  We'll call this the "first team".

* Click the "Scoreboard" link at the top of the page.  This should display a scoreboard showing all teams "tied for 1st" with no solutions and no penalty points -- the same as what is shown on the PC2 Scoreboard GUI.

* Repeat the above for a second team (start another browser session, log in another team, go to the scoreboard page).

* On the FIRST team's browser, click "Runs", then "Submit Problem". Use the resulting pop-up dialog to submit a correct solution for the first problem (if you've configured using the "sumithello" sample contest, this will be "hello").

* Verify the following:
  * The submitting team gets back a "Yes" popup notification for the submission.
  * The PC2 Scoreboard shows the "yes" for the submitting (first) team, and now has the first team in 1st place and all other teams tied for second place.
  * On the SECOND team's browser screen (which should have been displaying the **scoreboard** at the moment the first team submitted their run) the scoreboard automatically updated to show the "Yes" for the first team and now shows all other teams in second place.
  * When you click on the "Scoreboard" link on the first team's browser screen, it also shows the "Yes" for that team.

* On the PC2 Admin, make a change which could potentially alter the standings display.  One simple example would be to select some team account (say, "team 8") and mark them as "NOT shown on scoreboard".  Another example would be to edit the submission from the first team to have a different submission time.

* Verify that as soon as the above change is entered, the scoreboards on both the first and second team browsers update accordingly (for example, if say team 8 was marked "NOT shown on scoreboard, that team should immediately disappear from BOTH team browser displays).

* For further testing, submit runs for other problems and/or from other teams; all team browsers should at all times show the correct standings.
